<?xml version="1.0"?>
<launch>
    <!-- Arguments -->
    <arg name="model"            default="$(find xacro)/xacro '$(find igvc2023)/urdf/orange.xacro'" />
    <arg name="calibration"      default="$(find igvc2023)/params/VLP16db.yaml"/>
    <arg name="scan_dev"         default="/dev/sensors/hokuyo_urg"/>
    <arg name="imu_dev"          default="/dev/sensors/imu"/>
    <arg name="camera_dev"       default="/dev/sensors/insta360_air"/>
    <arg name="LED_dev"          default="/dev/sensors/LED"/>
    <arg name="gps_dev"          default="/dev/ttyACM0"/>
    <arg name="gps1_dev"         default="/dev/ttyACM1"/>
    <arg name="gps2_dev"         default="/dev/ttyACM2"/>
    <arg name="use_ekf"          default="true"/>
    <arg name="urg_ip"           default="192.168.3.11"/>
    
    
    <!-- Run a python script to the send a service call to gazebo_ros to spawn a URDF robot -->
    <param name="robot_description" command="$(arg model)"/>
    
    <!-- Convert joint states to TF transforms for rviz, etc -->
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" respawn="false" output="screen"/>
    
    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher">
        <param name="use_gui" value="true"/>
        <param name="rate"    value="50"/>
    </node>
    
    
    <!-- Node to setting up Arduino serial communication for IMU sensor -->
    <node pkg="rosserial_python" type="serial_node.py" name="serial_node_IMU">
        <param name="port" value="$(arg imu_dev)"/>
        <param name="baud" value="115200"/>
    </node>
    
    
    <!-- Node for setting up 2D-LiDAR via Ethernet connection -->
    <node pkg="urg_node" type="urg_node" name="urg_node">
        <param name="frame_id"                  value="/hokuyo_link"/>
        <param name="ip_adress"                 value="$(arg urg_ip)"/>
        <param name="angle_min"   type="double" value="-1.22"/>
        <param name="angle_max"   type="double" value="1.22"/>
        <remap from="/scan"       to="/hokuyo_scan"/>
    </node>
    
    
    <!-- Run velodyne_pointcloud/TransformNodelet in a nodelet manager for a VLP-16 -->
    <!-- Start nodelet manager -->
    <node pkg="nodelet" type="nodelet" name="velodyne_link_nodelet_manager" args="manager" output="screen"/>
    
    
    <!-- Load driver nodelet into it -->
    <node pkg="nodelet" type="nodelet" name="velodyne_link_nodelet_manager_driver" args="load velodyne_driver/DriverNodelet velodyne_link_nodelet_manager">
        <param name="device_ip"              value=""/>
        <param name="frame_id"               value="velodyne_link"/>
        <param name="model"                  value="VLP16"/>
        <param name="pcap"                   value=""/>
        <param name="port"                   value="2368"/>
        <param name="read_fast"              value="false"/>
        <param name="read_once"              value="false"/>
        <param name="repeat_delay"           value="0.0"/>
        <param name="rpm"                    value="600.0"/>
        <param name="gps_time"               value="false"/>
        <param name="pcap_time"              value="false"/>
        <param name="cut_angle"              value="-0.01"/>
        <param name="timestamp_first_packet" value="false"/>
    </node>
    
    
    <!-- Start transform nodelet -->
    <node pkg="nodelet" type="nodelet" name="velodyne_link_nodelet_manager_transform" args="load velodyne_pointcloud/TransformNodelet velodyne_link_nodelet_manager">
        <param name="model"          value="VLP16"/>
        <param name="calibration"    value="$(arg calibration)"/>
        <param name="fixed_frame"    value=""/>
        <param name="target_frame"   value=""/>
        <param name="max_range"      value="130.0"/>
        <param name="min_range"      value="0.4"/>
        <param name="organize_cloud" value="false"/>
    </node>
    
    
    <!-- Ground segmentation, Pointcloud to laserscan, Merge scans -->
    <include file="$(find igvc2023)/launch/segmentation.launch"/>
    
    <!-- Start camera -->
    <node pkg="cv_camera" type="cv_camera_node" name="cv_camera">
       <param name="device_path" value="/dev/sensors/insta360_air"/>
       <param name="image_width" value="1478"/>
       <param name="image_height" value="736"/>
    </node>

    <!-- Line detection node -->
    <include file="$(find line_detection)/launch/igvc2022_line_detection.launch">
        <arg name="is_sim" value="false"/>
    </include>
    
    <!-- Sensor fusion -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization" clear_params="true" output="screen">
        <rosparam command="load" file="$(find igvc2023)/config/ekf.yaml" />
        <remap from="odom0" to="/odom" />
        <remap from="odom1" to="/odometry/gps" />
        <remap from="imu0" to="/imu" />
    </node>
    
    <node pkg="robot_localization" type="navsat_transform_node" name="navsat_transform" clear_params="true" output="screen">
        <rosparam command="load" file="$(find igvc2023)/config/navsatekf.yaml" />
        <remap from="/imu/data" to="/imu" />
        <remap from="/gps/fix" to="/fix" />
    </node>
    
    <node pkg="nmea_navsat_driver" type="nmea_serial_driver" name="navsat" respawn="true">
        <param name="port" value="$(arg gps_dev)"/>
        <param name="baud" value="19200"/>
    </node>
    
</launch>
